src/leo_cache_tran.erl

```diff
@@ -0,0 +1,198 @@
+            {error, timeout};
+        {'EXIT', Reason} ->
+%%            gen_server:call(?MODULE, {unregister, Tbl, Key}, ?TRAN_TIMEOUT),
+            {error, Reason};
+        Ret ->
+            Ret
+    end.
+
+%% @doc
+-spec(done_tran(atom(), binary())->
+            ok).
+done_tran(Tbl, Key) ->
+    gen_server:call(?MODULE, {done_tran, Tbl, Key}, ?TRAN_TIMEOUT).
+
+
+%%--------------------------------------------------------------------
+%% GEN_SERVER CALLBACKS
+%%--------------------------------------------------------------------
+%% Function: init(Args) -> {ok, State}          |
+%%                         {ok, State, Timeout} |
+%%                         ignore               |
+%%                         {stop, Reason}
+%% Description: Initiates the server
+init([]) ->
+    ets:new(?PROCESSDB, [named_table, set, private]),
+    ets:new(?REPLYDB, [named_table, set, private]),
+    {ok, unused, ?TRAN_TIMEOUT}.
+
+handle_call(stop, _From, State) ->
+    {stop, normal, ok, State};
+
+%%handle_call({unregister, Tbl, Key}, From, State) ->
+%%    {Pid, _Ref} = From,
+%%    case ets:lookup(?REPLYDB, {Tbl, Key}) of
+%%        [{{Tbl, Key}, ReplyList}] ->
+%%            ?debugVal(ReplyList),
+%%            ReplyList2 = lists:filter(fun({Pid2, _Ref2}) ->
+%%                                              Pid2 =/= Pid
+%%                                      end, ReplyList),
+%%            ?debugVal(ReplyList2),
+%%            ets:insert(?REPLYDB, {{Tbl, Key}, ReplyList2});
+%%        _ ->
+%%            void
+%%    end,
+%%    {reply, ok, State, ?TRAN_TIMEOUT};
+
+handle_call({tran, Pid, Tbl, Key}, _From, State) ->
+    MonitorRef = erlang:monitor(process, Pid),
+    ets:insert(?PROCESSDB, {MonitorRef, {Tbl, Key}}), 
+    ets:insert_new(?REPLYDB, {{Tbl, Key}, []}),
+    {reply, ok, State, ?TRAN_TIMEOUT}; 
+
+handle_call({has_tran, Tbl, Key}, From, State) ->
+    case ets:lookup(?REPLYDB, {Tbl, Key}) of
+        [{{Tbl, Key}, ReplyList}] ->
+            ets:insert(?REPLYDB, {{Tbl, Key}, [From | ReplyList]}),
+            {noreply, State, ?TRAN_TIMEOUT};
+        _ ->
+            {reply, {ok, not_found}, State, ?TRAN_TIMEOUT}
+    end;
+
+handle_call({done_tran, Tbl, Key}, _From, State) ->
+    reply_all(Tbl, Key),
+    {reply, ok, State, ?TRAN_TIMEOUT}.
+
+handle_cast(_Msg, State) ->
+    {noreply, State, ?TRAN_TIMEOUT}.
+
+
+%% Function: handle_info(Info, State) -> {noreply, State}          |
+%%                                       {noreply, State, Timeout} |
+%%                                       {stop, Reason, State}
+%% Description: Handling all non call/cast messages
+%% handle_info({_Label, {_From, MRef}, get_modules}, State) ->
+%%     {noreply, State};
+handle_info({'DOWN', MonitorRef, _Type, _Pid,_Info}, State) ->
+    case ets:lookup(?PROCESSDB, MonitorRef) of
+        [{MonitorRef, {Tbl, Key}}] ->
+            reply_all(Tbl, Key);
+        _ ->
+            void
+    end,
+    ets:delete(?PROCESSDB, MonitorRef),
```

Comment:
I think I have incorrect understanding of the leo_gateway processes flow then.
In such case, leo_gateway need to explicitly call `leo_cache_tran:done_tran/2` to release the lock.

About deleting record in ?PROCESSDB, the module probably need another mapping of `Pid -> MonitorRef` or the reverse mapping of `{type, key} -> MonitorRef` then. Otherwise I guess the whole table need to be transversed once.